### ISSUE TITLE
Support new auth combinators

### ISSUE THREAD

**Issue by 3noch** [imalsogreg/servant-reflex#5]
It doesn't look like the new auth combinators are supported yet.

**Comment by 3noch** [comment]
For `BasicAuth` I think it'd be something like this, but I can't seem to get the types right. I'm not sure how to lift things into the Event framework.

```haskell
instance (HasClient t m api, Reflex t) => HasClient t m (BasicAuth realm usr :> api) where
  type Client t m (BasicAuth realm usr :> api) = Behavior t (Maybe BasicAuthData) -> Client t m api

  clientWithRoute Proxy q req baseurl (BasicAuthData username password) =
    clientWithRoute (Proxy :: Proxy api) q req' baseurl
      where
        req'    = req { _xhrRequest_config = config' }
        config' = _xhrRequest_config req
                { _xhrRequestConfig_user     = Just username
                , _xhrRequestConfig_password = Just password
                }
```

**Comment by imalsogreg** [comment]
Sorry for the delay. I think your type for Client is right (s / Maybe / Either String ).

The `req` you're record-updating isn't a `XhrRequestConfig` though, it's a `Req t`. So what we would do is add a field to the `Req t` type, called `reqBasicAuth :: Maybe (Behavior t (Either String BasicAuthData))`. Then then we'd handle merging that field in with all the other `Req t` fields in the `Common.Req` module.

Do you feel like picking this back up, or should I work on it? I'd definitely like to have it. I need to add the Auth stuff to servant-snap too. Oh - until servant-snap can handle auth combinators, I may have to rely on a servant-server (wai) test server.

**Comment by 3noch** [comment]
Thanks for the feedback! Unfortunately I got stymied due to lots of travel and won't be able to work on anything for a couple weeks. So that's the earliest I could work on it. If it's still open by then I'll give it a whack.

**Comment by imalsogreg** [comment]
👍

**Comment by lethom** [comment]
I will try to have a look at this this weekend. I really think this is a must-have for servant-reflex to be usable.

**Comment by 3noch** [comment]
I'm back! A lot of stuff changed since a over a year ago and I'm finally in need of this yet again! I'm hoping to use [servant-auth](https://hackage.haskell.org/package/servant-auth). What's the story for auth using servant-reflex at this point?

**Comment by imalsogreg** [comment]
@3noch Excellent! For my Auth needs I've been relying on a combination of setting cookie headers (on the server side), and ignoring auth on the client side (letting it be handled entirely by the cookies). I'm glad to see that `servant-auth` has an extremely light dependency set, so there's probably no reason for `servant-reflex` to avoid adding it as a dependency and implementing the XHR-twiddling. Any interest in working on a PR? If not, I may be able to get to in in the next couple of weeks.

**Comment by 3noch** [comment]
@imalsogreg I'll add a PR ASAP but I may need some help!

**Comment by imalsogreg** [comment]
Closed by @3noch himself #36. Thank you!!
